### DISCUSSION TITLE
Ping/Connect to Inside Tunnel IPs?

### DISCUSSION THREAD

**Discussion by coofercat** [strongswan/strongswan#647] [Q&A]
I have Strongswan running on Ubuntu, connecting to AWS VPN. I have had this working with a "static" VPN, so I know my ipsec is good. However, I'd like to use BGP, which in AWS parlance means a "dynamic" VPN.

In this mode, the VPN ipsec is established the same way. Then, the AWS-side inside tunnel IP becomes a BGP neighbor which tries to talk to the Strongswan-side inside tunnel IP, where it expects to find a BGP server. Once connected, they negotiate and then the VPN starts to work.

I found an article describing exactly this setup here: https://www.edge-cloud.net/2019/07/18/aws-site-2-site-vpn-with-strongswan-frrouting/#strongswan-setup

However, in my case, I have my ipsec tunnel up, I have vti tunnel devices up, and if I `tcpdump` on the tunnel devices, I can see the AWS device sending TCP SYN packets to my BGP server. I can see a TCP ACK in the tcpdump output, but this never seems to reach AWS, and so the connection is never established:

```
17:20:35.204767 IP 169.254.140.145.32827 > 169.254.140.146.179: Flags [S], seq 451362349, win 26880, options [mss 1375,sackOK,TS val 3364049608 ecr 0,nop,wscale 7], length 0
17:20:35.204814 IP 169.254.140.146.179 > 169.254.140.145.32827: Flags [S.], seq 3764245674, ack 451362350, win 64249, options [mss 1379,sackOK,TS val 4115933632 ecr 3364049608,nop,wscale 7], length 0
```
Likewise, if I try to `ping` the AWS inside tunnel IP, this also doesn't seem to actually reach AWS - I don't get a reply, but also I don't see the AWS Cloudwatch metrics for "bytes in/out" on the VPN going up. I also don't see additional encrypted traffic leaving my Strongswan server. So I'm pretty confident that even though the packets are on my vti device, they're not getting sent "up" the tunnel to AWS.

AWS usually makes two tunnels - for now I'm just using one (when I get that working, I'll add the second).

The details of all this are:
My local network CIDR: 10.0.16.0/20
AWS network CIDR: 10.0.32.0/20
Inside tunnel AWS side: 169.254.140.145/30
Inside tunnel Strongswan side: 169.254.140.146/30
Strongswan public IP 1.1.1.1
AWS public IP: 2.2.2.2

My ipsec.conf:
```
config setup
  uniqueids=no

conn %default
  leftauth=psk
  rightauth=psk
  ike=aes128-sha1-modp1024!
  ikelifetime=28800s
  aggressive=no
  esp=aes128-sha1-modp1024!
  lifetime=3600s
  type=tunnel
  dpddelay=10s
  dpdtimeout=30s
  keyexchange=ikev1
  rekey=yes
  reauth=no
  dpdaction=restart
  closeaction=restart
  leftsubnet=0.0.0.0/0,::/0
  rightsubnet=0.0.0.0/0,::/0
  leftupdown=/etc/ipsec-vti.sh
  installpolicy=yes
  compress=no
  mobike=no

conn AWS-GW1
  auto=start
  left=1.1.1.1
  leftid=1.1.1.1
  right=2.2.2.2
  ike=aes128-sha1-modp1024
  ikelifetime=8h
  esp=aes128-sha1-modp1024
  keyingtries=%forever
  forceencaps=yes
  mark=100
```
My `ip a` for tunnel1 looks like this:
```
48: tunnel1@NONE: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1419 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ipip 1.1.1.1 peer 2.2.2.2
    inet 169.254.140.146 peer 169.254.140.145/30 scope global tunnel1
       valid_lft forever preferred_lft forever
    inet6 fe80::200:5efe:c6f4:8f78/64 scope link
       valid_lft forever preferred_lft forever
```
My `ip routes` look like this:
```
10.0.16.0/24 dev ens4 proto kernel scope link src 10.0.16.170
10.0.32.0/20 dev tunnel1 scope link src 10.0.16.170 metric 100
169.254.140.144/30 dev tunnel1 proto kernel scope link src 169.254.140.146
169.254.169.254 via 10.0.16.51 dev ens4
1.1.1.1 dev ens3 scope link
```

My `ip xfrm policy list` looks like this:
```
src 10.0.16.0/20 dst 10.0.32.0/20
	dir out priority 379519
	mark 0x64/0xffffffff
	tmpl src 1.1.1.1 dst 2.2.2.2
		proto esp spi 0xc118c621 reqid 2 mode tunnel
src 10.0.32.0/20 dst 10.0.16.0/20
	dir fwd priority 379519
	mark 0x64/0xffffffff
	tmpl src 2.2.2.2 dst 1.1.1.1
		proto esp reqid 2 mode tunnel
src 10.0.32.0/20 dst 10.0.16.0/20
	dir in priority 379519
	mark 0x64/0xffffffff
	tmpl src 2.2.2.2 dst 1.1.1.1
		proto esp reqid 2 mode tunnel
```

This feels like I'm possibly missing some routes, but maybe am missing some xfrm policy to make 169.254.140.144/30 traffic "go up" the tunnel, which it seems it's not at the moment - although I can't figure out what/where it's being dropped  (the iptables firewall is wide open right now).

Anyone got any ideas? Anyone ever got a VPN to AWS working and can try pinging the inside IPs? Or has anyone got an AWS BGP-based VPN working? Any help would be much appreciated :-)

**Comment by coofercat** [comment]
It turns out the solution is an additional xfrm policy. The thinking behind it is that whilst packets are getting onto the interface, the traffic isn't being "scooped up" by ipsec. To resolve the problem, it seems we need to mark the packets (in the same way as we do in the ipsec config and iptables), and also to specifically "activate" xfrm by policy. In my case, it was this:

```
ip xfrm policy add dst 169.254.140.144/30 src 169.254.140.144/30 dir out tmpl src 1.1.1.1 dst 2.2.2.2 proto esp spi 0xc0f93fba reqid 1 mode tunnel mark 0x64
```
This makes a policy that looks like this:
```
src 169.254.140.144/30 dst 169.254.140.144/30
	dir out priority 0
	mark 0x64/0xffffffff
	tmpl src 1.1.1.1 dst 2.2.2.2
		proto esp spi 0xc0f93fba reqid 1 mode tunnel
```
This contains some magic, which it seems we have to get from the existing policy. The first thing is the "spi number". This gets set by Strongswan, and isn't, I don't think, predictable, nor available in the environment passed to `ipsec-vti.sh`. Instead, you'll need to do some grep and awk to pull it out of `ip xfrm policy` output. Likewise the `reqid` - this set with 1 for the first interface and 2 for the second that gets created - sadly, tunnel1 isn't always the first, so you'll need to pull that number out too.

**Reply by tobiasbrunner** [reply]
While you could maybe "fix" the issue with additional policies (the other peer probably has to do the same, unless it accepts just any traffic, and _fwd_/_in_ policies might be required to), the question is why your peer narrowed the traffic selectors to `10.0.16.0/20 === 10.0.32.0/20` in the first place when the goal is to tunnel between `169.254.140.144/30 === 169.254.140.144/30` too. You configured `0.0.0.0/0,::/0` as traffic selector, so why does the peer not do the same?

Also you technically don't need the SPI in the _out_ policy (we just set it so we can switch outbound SAs more predictably during rekeying). And the reqid is passed in `PLUTO_REQID` to the updown script.

**Reply by coofercat** [reply]
Interesting - I wondered why this network was "excluded" too. I haven't found a solid source, but at least one article somewhere on the Internet suggested that once their tunnels were up that they could ping the inside addresses - if true, it would seem their policies were created correctly.

FWIW, if I build the VPN on the AWS side as a "static" VPN, then there's no BGP requirement, so the tunnels show as "up" in the AWS console as soon as Ipsec is up. In such a state, the inside addresses aren't pingable in my setup either, so it's not likely anything to do with AWS or the ipsec setup.

If I understand you correctly, what we could expect would be xfrm policies that look more like this:
```
src 0.0.0.0/0 dst 0.0.0.0/0
	dir out priority 0
	mark 0x64/0xffffffff
	tmpl src 1.1.1.1 dst 2.2.2.2
		proto esp spi 0xc0f93fba reqid 1 mode tunnel
```
I didn't try this (as I was trying to incrementally fix what was there), but I could imagine this working fine, and far more simply than what's actually being created.

Curiously, the config for 10.0.32.0/20 is only in the `ipsec-vti.sh` script (and then only used to insert some `ip route` rules, which happen long after the xfrm policies are created). The local CIDR (10.0.16.0/20) is __only__ in the BGP configs (frr.conf) - even the local interface has a slightly different setting (it's got a /24 netmask). These configs then, aren't anywhere I'd expect, therefore, I'm not even sure _how_ it's getting configured into the xfrm policies (is it something that's coming from the AWS end?). Either way, PLUTO_REQID works nicely - thanks for that.

**Reply by tobiasbrunner** [reply]
The traffic selectors on which the policies are based get negotiated when the CHILD_SA is created, so check the log. Most likely the other end has some specific subnets configured and your proposed traffic selectors get narrowed, so make sure to also check the corresponding tunnel config there.

**Reply by coofercat** [reply]
Thanks @tobiasbrunner for your help here - I've finally been able to try your ideas in some different configs, and have written it up as a separate answer (which is far more "correct" than the one I started with!).

**Comment by coofercat** [comment]
[Making this an answer instead of a reply, as it's hopefully more useful that way]

The solution is *not* to add more xfrm policies. The problem with that approach seems to be that after an hour, ipsec renegotiates and loses the additional policies. I originally thought Strongswan would "down" and "up" the tunnel for this, but it does not - after all, Ipsec is up and working, so actually it's reasonable behaviour to leave the tunnel in place (albeit in state that doesn't work because BGP can't negotiate so will remove routes to it).

The full solution is to ensure both ends of the ipsec tunnel use 0.0.0.0/0 as their traffic selectors. This makes the xfrm policies created by Strongswan work immediately, without needing to 'hack' anything. Right away, access to the remote network and the inside-tunnel IPs is possible by a single policy in and a single policy out (which is automatically created and maintained). This approach also works across Ipsec renegotiations, so no need to continuously monitor and restart strongswan or whatever else (well, not for this reason, at least).

Lastly, a note to AWS users. If you use 'static' VPNs (without BGP), then the traffic selectors you use there (local_ipv4_network_cidr and remote_ipv4_network_cidr in AWs parlance) has an effect on the network routes that AWS attaches to the VPN and VPC connected to it. If however you use 'dynamic' VPNs (with BGP), then the traffic selectors do not affect routes at all - only the routes propagated to the VPN (from your VPC, Transit Gateway or whatever) have any effect. So for dynamic VPNs, always use 0.0.0.0/0 as the traffic selectors.

**Reply by tobiasbrunner** [reply]
> The problem with that approach seems to be that after an hour, ipsec renegotiates and loses the additional policies.

That doesn't really make sense.  Policies don't just disappear (unlike e.g. routes if an interface goes away). Did you check the log (and maybe `ip xfrm policy`) to see what actually happens?

**Reply by coofercat** [reply]
Actually, I misspoke - I should have said "it messes up the policies" ;-) I tried to get a nice log entry about this, but honestly, never found anything that I thought to be useful. I'm afraid I didn't keep it, but the xfrm policies seemed to change to something completely different - my guess is that my inserted policies no longer matched up with the new ones (presumably the `spi 0xc0f93fba` no longer matches). The point being that the `ipsec-vti.sh` script wasn't used to re-setup the system after the ipsec renegotiation, so my interventions weren't taken into account.

**Reply by tobiasbrunner** [reply]
I see. Yeah, I mentioned that the SPI isn't required (and why we set it) above, but I guess I didn't explain exactly why it isn't a good idea to use it in an updown script (which, as you noticed, isn't called during CHILD_SA rekeyings). Anyway, avoiding all that is definitely the better approach.
